docs/user-guide/kubectl-conventions.md

```diff
@@ -8,15 +8,45 @@ assignees:
 * TOC
 {:toc}
 
-## Using `kubectl` in Reusable Scripts 
+## Using `kubectl` in Reusable Scripts
 
 If you need stable output in a script, you should:
 
-* Request one of the machine-oriented output forms, such as `-o name`, `-o json`, `-o yaml`, `-o go-template`, or `-o jsonpath` 
+* Request one of the machine-oriented output forms, such as `-o name`, `-o json`, `-o yaml`, `-o go-template`, or `-o jsonpath`
 * Specify `--output-version`, since those output forms (other than `-o name`) output the resource using a particular API version
 * Specify `--generator` to pin to a specific behavior forever, if using generator-based commands (such as `kubectl run` or `kubectl expose`)
 * Don't rely on context, preferences, or other implicit state
 
+## Generators
+
+`kubectl run` provides following generators:
+
+* `run/v1` - creates a replication controller.
+* `run-pod/v1` - creates a pod.
+* `deployment/v1beta1` - creates a deployment.
+* `job/v1beta1` - creates a job (using `extensions/v1beta1' endpoint).
+* `job/v1` - creates a job.
+* `scheduledjob/v2alpha1` - creates a scheduled job.
+
+Following rules are implemented if no generator is specified for cluster version at least 1.3:
+
+* For `--restart=Always' a deployment generator (`deployment/v1beta1`) is chosen.
+* For `--restart=OnFailure` a job generator (`job/v1`) is chosen.
+* For `--restart=Never` a pod generator (`run-pod/v1`) is chosen.
+
+Additionally, for cluster version at least 1.4 if `--schedule` flag is set a scheduled job
+generator ('scheduledjob/v2alpha1') is chosen.
+
+For cluster version 1.2:
+
+* For `--restart=Always` a deployment generator ('deployment/v1beta1') is chosen.
+* For other restart policy a job generator ('job/v1') is chosen.
+
+For cluster version 1.1 and older:
+
+* For `--restart=Always` a replication controller ('run/v1') is chosen
+* For other restart policy a pod generator ('run-pod/v1') is chosen.
```

Comment:
Perhaps something like this is easier to read?

|   Generated Resource   | Cluster 1.4         | Cluster 1.3                 | Cluster 1.2                                 | Cluster <=1.1                               |
|:----------------------:|---------------------|---------------------|-------------------------------------|-------------------------------------|
| Scheduled Job          | `--schedule=<cron>`          | N/A                 | N/A                                 | N/A                                 |
| Deployment             | `--restart=Always`    | `--restart=Always`    | `--restart=Always`                    | N/A                                 |
| Replication Controller | `--generator=run/v1`  | `--generator=run/v1`  | `--generator=run/v1`                  | `--restart=Always`                    |
| Job                    | `--restart=OnFailure` | `--restart=OnFailure` | `--restart=OnFailure` OR `--restart=Never` | N/A                                 |
| Pod                    | `--restart=Never`     | `--restart=Never`     | `--generator=run-pod/v1`              | `--restart=OnFailure` OR `--restart=Never` |

This isn't clean enough though, we should also mention that `--schedule` sets the `--generator` for you, as well as `--restart`, and if you set the generator to something else it wouldn't work as expected. Also, once `--schedule` set the `--generator`, `--restart` won't set it again (order: `--schedule` --> `--restart` --> `--generator`).